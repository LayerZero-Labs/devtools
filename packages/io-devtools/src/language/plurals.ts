const cardinalRules = new Intl.PluralRules('en-US')

const ordinalRules = new Intl.PluralRules('en-US', { type: 'ordinal' })
const ordinals: Record<Intl.LDMLPluralRule, string> = {
    one: 'st',
    two: 'nd',
    few: 'rd',
    other: 'th',
    zero: 'th',
    many: 'th',
}

/**
 * Turn a number into an ordinal.
 *
 * ```typescript
 * pluralizeOrdinal(7) // 7th
 * pluralizeOrdinal(1) // 1st
 * pluralizeOrdinal(19) // 19th
 * ```
 *
 * @param {number} n
 * @returns {string}
 */
export const pluralizeOrdinal = (n: number): string => {
    const rule = ordinalRules.select(n)
    const suffix = ordinals[rule]

    return `${n}${suffix}`
}

/**
 * Choose a correct form of a noun based on cardinality.
 *
 * ```typescript
 * pluralizeNoun(7, 'cat') // cats
 * pluralizeNoun(1, 'cat') // cat
 * pluralizeNoun(19, 'cactus', 'cacti') // cacti
 * ```
 *
 * @param {number} n
 * @param {string} singular The signular form of the english noun
 * @param {string} [plural] Plural version of the noun for irregular cases
 * @returns {string}
 */
export const pluralizeNoun = (n: number, singular: string, plural: string = `${singular}s`): string => {
    const rule = cardinalRules.select(n)
    if (rule === 'one') {
        return singular
    }

    return plural
}
