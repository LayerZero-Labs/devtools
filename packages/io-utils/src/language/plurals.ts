const cardinalRules = new Intl.PluralRules('en-US')

const ordinalRules = new Intl.PluralRules('en-US', { type: 'ordinal' })
const ordinals: Record<Intl.LDMLPluralRule, string> = {
    one: 'st',
    two: 'nd',
    few: 'rd',
    other: 'th',
    zero: 'th',
    many: 'th',
}

/**
 * Turn a number into an ordinal.
 *
 * ```typescript
 * pluralizeOrdinal(7) // 7th
 * pluralizeNoun(1) // 1st
 * pluralizeNoun(19, 'cacti') // 19th
 * ```
 *
 * @param {number} n
 * @returns {string}
 */
export const pluralizeOrdinal = (n: number): string => {
    const rule = ordinalRules.select(n)
    const suffix = ordinals[rule]

    return `${n}${suffix}`
}

/**
 * Choose a correct form of a noun based on cardinality.
 *
 * ```typescript
 * pluralizeNoun('cat', 7) // cats
 * pluralizeNoun('cat', 1) // cat
 * pluralizeNoun('cactus', 19, 'cacti') // cacti
 * ```
 *
 * @param {string} singular The signular form of the english noun
 * @param {number} n
 * @param {string} [plural] Plural version of the noun for irregular cases
 * @returns {string}
 */
export const pluralizeNoun = (singular: string, n: number, plural: string = `${singular}s`): string => {
    const rule = cardinalRules.select(n)
    if (rule === 'one') return singular

    return plural
}
