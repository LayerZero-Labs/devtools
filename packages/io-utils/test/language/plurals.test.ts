import { pluralizeNoun, pluralizeOrdinal } from '@/language'

describe('language/plurals', () => {
    describe('pluralizeOrdinal', () => {
        it.each([0, 1, 2, 3, 4, 5, 11, 12, 21, 100, 1234])(`should work for %d`, (n) =>
            expect(pluralizeOrdinal(n)).toMatchSnapshot()
        )
    })

    describe('pluralizeNoun', () => {
        describe('without custom plural', () => {
            it.each([0, 1, 2, 3, 4, 5, 11, 12, 21, 100, 1234])(`should work for %d`, (n) =>
                expect(pluralizeNoun(n, 'cat')).toMatchSnapshot()
            )
        })

        describe('with custom plural', () => {
            it.each([0, 1, 2, 3, 4, 5, 11, 12, 21, 100, 1234])(`should work for %d`, (n) =>
                expect(pluralizeNoun(n, 'cactus', 'cacti')).toMatchSnapshot()
            )
        })
    })
})
