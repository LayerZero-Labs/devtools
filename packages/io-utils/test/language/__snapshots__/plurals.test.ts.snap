// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`language/plurals pluralizeNoun with custom plural should work for 0 1`] = `"cacti"`;

exports[`language/plurals pluralizeNoun with custom plural should work for 1 1`] = `"cactus"`;

exports[`language/plurals pluralizeNoun with custom plural should work for 2 1`] = `"cacti"`;

exports[`language/plurals pluralizeNoun with custom plural should work for 3 1`] = `"cacti"`;

exports[`language/plurals pluralizeNoun with custom plural should work for 4 1`] = `"cacti"`;

exports[`language/plurals pluralizeNoun with custom plural should work for 5 1`] = `"cacti"`;

exports[`language/plurals pluralizeNoun with custom plural should work for 11 1`] = `"cacti"`;

exports[`language/plurals pluralizeNoun with custom plural should work for 12 1`] = `"cacti"`;

exports[`language/plurals pluralizeNoun with custom plural should work for 21 1`] = `"cacti"`;

exports[`language/plurals pluralizeNoun with custom plural should work for 100 1`] = `"cacti"`;

exports[`language/plurals pluralizeNoun with custom plural should work for 1234 1`] = `"cacti"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 0 1`] = `"cats"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 1 1`] = `"cat"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 2 1`] = `"cats"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 3 1`] = `"cats"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 4 1`] = `"cats"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 5 1`] = `"cats"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 11 1`] = `"cats"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 12 1`] = `"cats"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 21 1`] = `"cats"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 100 1`] = `"cats"`;

exports[`language/plurals pluralizeNoun without custom plural should work for 1234 1`] = `"cats"`;

exports[`language/plurals pluralizeOrdinal should work for 0 1`] = `"0th"`;

exports[`language/plurals pluralizeOrdinal should work for 1 1`] = `"1st"`;

exports[`language/plurals pluralizeOrdinal should work for 2 1`] = `"2nd"`;

exports[`language/plurals pluralizeOrdinal should work for 3 1`] = `"3rd"`;

exports[`language/plurals pluralizeOrdinal should work for 4 1`] = `"4th"`;

exports[`language/plurals pluralizeOrdinal should work for 5 1`] = `"5th"`;

exports[`language/plurals pluralizeOrdinal should work for 11 1`] = `"11th"`;

exports[`language/plurals pluralizeOrdinal should work for 12 1`] = `"12th"`;

exports[`language/plurals pluralizeOrdinal should work for 21 1`] = `"21st"`;

exports[`language/plurals pluralizeOrdinal should work for 100 1`] = `"100th"`;

exports[`language/plurals pluralizeOrdinal should work for 1234 1`] = `"1234th"`;
